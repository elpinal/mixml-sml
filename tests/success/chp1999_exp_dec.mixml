link X =
  link E = {module Expr = {type exp : Type}} with
  {module Decl =
    link X = {type dec : Type} with
    {val make_val : int -> ext E.Expr.exp -> ext X.dec}
  }
with

link
{module Expr =
  link S = {type u = [Let : ext X.Decl.dec * ext X.Expr.exp]} with
  link Y = new [unit {data exp = ext S.u}] with
  link {val make_let = fun (d : ext X.Decl.dec) (e : ext Y.exp) -> ext Y.exp_in (Let(d, e) : ext S.u)} with
  {val make_let_val = fun (id : int) (e1 : ext Y.exp) (e2 : ext Y.exp) ->
    ext Y.exp_in (Let(ext X.Decl.make_val id e1, e2) : ext S.u)
  }
} with

{module Decl =
  link S = {type u = [Val : int * ext X.Expr.exp]} with
  link Y = {data dec = ext S.u} with
  {val make_val = fun (n : int) (e : ext X.Expr.exp) -> ext Y.dec_in (Val(n, e) : ext S.u)}
}
