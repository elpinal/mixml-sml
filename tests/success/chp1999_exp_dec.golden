{Decl : {dec : [= !1], dec_in : [[Val : int * !0] -> !1]+, dec_out : [!1 -> [Val : int * !0]]+, make_val : [int -> !0 -> !1]+, u : [= [Val : int * !0]]}, Expr : {exp : [= !0], exp_in : [[Let : !1 * !0] -> !0]+, exp_out : [!0 -> [Let : !1 * !0]]+, make_let : [!1 -> !0 -> !0]+, make_let_val : [int -> !0 -> !0 -> !0]+, u : [= [Let : !1 * !0]]}}
